### PR TITLE
fix application behaviour hurting OTP's "release_handler"

### PR DESCRIPTION
**Pull request by fdie**
Hi, 

Many thanks for your open source.
We just noticed that your application breaks the "release_handler" of OTP.

The child <0.9438.0> of the application_master's child is not a supervisor process.

```erl
application_controller:get_master(kafka_protocol).
<0.11604.0>

application_master:get_child(<0.11604.0>).        
{<0.9438.0>,kpro_app}

sys:get_state(<0.9438.0>).
** exception exit: {timeout,{sys,get_state,[<0.9438.0>]}}
     in function  sys:send_system_msg/2 (sys.erl, line 338)
     in call from sys:get_state/1 (sys.erl, line 139)
```

Thus any release upgrade breaks too :

```erl
release_handler:install_release("1.161.0").      
{error,{'EXIT',{timeout,{sys,get_status,[<0.9438.0>]}}}}
```

2022-04-12 15:15:47.788919-04:00 [error] <0.131.0> release_handler: cannot find top supervisor for application kafka_protocol

Best regards,
Frank

### CODE REVIEW

File: src/kpro_sup.erl [hunk 1]
```diff
@@ -0,0 +1,13 @@
+-module(kpro_sup).
+
+-behaviour(supervisor).
+
+-export([start_link/0, init/1]).
+
+start_link() -> supervisor:start_link({local, ?MODULE}, ?MODULE, []).
```
Comment (mikpe) [inline src/kpro_sup.erl]: If you do this you also need to add `?MODULE` to the `registered` list in `.app.src`.

Comment (fdie) [reply src/kpro_sup.erl]: You are right, its useless to register an unused supervisor, will remove the registration. Thx

**Comment by mikpe** [comment]
So the reason this issue crept up is because the change to allow user-provided compression modules wanted to migrate settings from app env to persistent term on init, which prevented kafka_protocol from remaining a library application. Oh well.

**Review by mikpe — APPROVED** [review]

**Comment by fdie** [comment]
Yes definitely, no issue before the "mod" in the "kafka_protocol.app.src" was specified.
Another solution could be, to remove the "kpro_app" module and do the init in the "kpro_compression" module like this: 

-on_load(init/0).
init() ->
  Provides = application:get_env(kafka_protocol, provide_compression, []),
  provide(Provides).

**Comment by fdie** [comment]
No, I'm wrong. It seems only apps support parameters like this
erl -pa _build/default/lib/*/ebin -kafka_protocol provide_compression "[{lz4, xxx}]"

Or you support extra parameters like that:
erl -pa _build/default/lib/*/ebin -extra provide_compression "[{lz4, xxx}]"
init:get_plain_arguments().
["provide_compression","[{lz4, lz}]"]

Or use another app ;-)
erl -pa _build/default/lib/*/ebin -kernel provide_compression "[{lz4, xxx}]"
application:get_env(kernel, provide_compression, []).
[{lz4, xxx}]

Or use env variables
erl -pa _build/default/lib/*/ebin -env provide_compression "[{lz4, lz}]"
os:getenv("provide_compression").
"[{lz4, lz}]"

**Comment by fdie** [comment]
Maybe you prefer to keep the "kpro_app" module to be compliant with your doc concerning configuration...
But the module's "start" method has to return the "top" supervisor PID as modified in this PR

**Comment by mikpe** [comment]
Ping @zmstone -- wdyt?

**Comment by zmstone** [comment]
An empty supervisor won’t hurt
I’d prefer to keep it simple. PR looks good, thank you for the pr and ideas.

**Comment by zmstone** [comment]
Please help to update changelog.md @fdie

**Comment by fdie** [comment]
Hi, as described here https://www.erlang.org/doc/design_principles/applications.html#application-callback-module

Changelog:

* The application ˋstartˋ method must return the pid of the top supervisor

**Comment by zmstone** [comment]
> Hi, as described here https://www.erlang.org/doc/design_principles/applications.html#application-callback-module
> 
> Changelog:
> 
> * The application ˋstartˋ method must return the pid of the top supervisor

thanks. But I mean update the file in this repo: `./changelog.md` (With a new patch release)

**Comment by fdie** [comment]
done ;-)

**Review by mikpe — APPROVED** [review]

**Comment by zmstone** [comment]
sorry for the delay, included in 4.1.0 release.

**Pull request merged**